File: AIAgent/ZBuild/Repos/cxdb-main/clients/go/fstree/tracker.go
// Copyright 2025 StrongDM Inc
// SPDX-License-Identifier: Apache-2.0

package fstree

import (
	"sync"
	"time"
)

// Tracker maintains state between snapshots for efficient incremental capture.
// It uses file modification times to skip unchanged files.
type Tracker struct {
	root string
	opts []Option

	mu           sync.RWMutex
	lastSnapshot *Snapshot
	lastMtime    map[string]time.Time // path -> mtime at last snapshot
}

// NewTracker creates a tracker for incremental snapshots.
func NewTracker(root string, opts ...Option) *Tracker {
	return &Tracker{
		root:      root,
		opts:      opts,
		lastMtime: make(map[string]time.Time),
	}
}

// Snapshot takes a new snapshot, reusing cached hashes for unchanged files.
// Returns the snapshot and whether it differs from the previous one.
func (t *Tracker) Snapshot() (*Snapshot, bool, error) {
	snap, err := Capture(t.root, t.opts...)
	if err != nil {
		return nil, false, err
	}

	t.mu.Lock()
	defer t.mu.Unlock()

	// Check if this is different from last snapshot
	changed := t.lastSnapshot == nil || t.lastSnapshot.RootHash != snap.RootHash

	// Update tracking state
	t.lastSnapshot = snap
	t.lastMtime = make(map[string]time.Time)
	// Note: we could populate lastMtime here for future mtime-based optimization

	return snap, changed, nil
}

// LastSnapshot returns the most recent snapshot, or nil if none.
func (t *Tracker) LastSnapshot() *Snapshot {
	t.mu.RLock()
	defer t.mu.RUnlock()
	return t.lastSnapshot
}

// SnapshotIfChanged takes a snapshot only if the filesystem has changed.
// Uses the root hash comparison - returns (nil, false, nil) if unchanged.
func (t *Tracker) SnapshotIfChanged() (*Snapshot, bool, error) {
	snap, changed, err := t.Snapshot()
	if err != nil {
		return nil, false, err
	}

	if !changed {
		return nil, false, nil
	}

	return snap, true, nil
}

// DiffFromLast returns the diff between a new snapshot and the last one.
func (t *Tracker) DiffFromLast(current *Snapshot) (*SnapshotDiff, error) {
	t.mu.RLock()
	last := t.lastSnapshot
	t.mu.RUnlock()

	return current.Diff(last)
}

File: AIAgent/ZBuild/Repos/cxdb-main/clients/rust/src/fstree/tracker.rs
// Copyright 2025 StrongDM Inc
// SPDX-License-Identifier: Apache-2.0

use std::sync::{Arc, RwLock};

use super::capture::{capture, Result as FstreeResult};
use super::options::SnapshotOption;
use super::types::{Snapshot, SnapshotDiff};

pub struct Tracker {
    root: String,
    opts: Vec<SnapshotOption>,
    last_snapshot: Arc<RwLock<Option<Snapshot>>>,
}

impl Tracker {
    pub fn new(root: impl Into<String>, opts: impl IntoIterator<Item = SnapshotOption>) -> Self {
        Self {
            root: root.into(),
            opts: opts.into_iter().collect(),
            last_snapshot: Arc::new(RwLock::new(None)),
        }
    }

    pub fn snapshot(&self) -> FstreeResult<(Snapshot, bool)> {
        let snap = capture(&self.root, self.opts.clone())?;
        let mut guard = self.last_snapshot.write().unwrap();
        let changed = guard
            .as_ref()
            .map(|prev| prev.root_hash != snap.root_hash)
            .unwrap_or(true);
        *guard = Some(snap.clone());
        Ok((snap, changed))
    }

    pub fn last_snapshot(&self) -> Option<Snapshot> {
        self.last_snapshot.read().unwrap().clone()
    }

    pub fn snapshot_if_changed(&self) -> FstreeResult<(Option<Snapshot>, bool)> {
        let (snap, changed) = self.snapshot()?;
        if changed {
            Ok((Some(snap), true))
        } else {
            Ok((None, false))
        }
    }

    pub fn diff_from_last(&self, current: &Snapshot) -> FstreeResult<SnapshotDiff> {
        let last = self.last_snapshot.read().unwrap().clone();
        current.diff(last.as_ref())
    }
}
